File: neoforge/src/main/java/wile/redstonepen/gametest/RelayGameTests.java
package wile.redstonepen.gametest;

import net.minecraft.gametest.framework.GameTest;
import net.neoforged.neoforge.gametest.GameTestHolder;
import net.neoforged.neoforge.gametest.PrefixGameTestTemplate;
import net.minecraft.gametest.framework.GameTestHelper;
import wile.redstonepen.ModConstants;
import wile.redstonepen.gametestcommon.RelayTests;

@GameTestHolder(ModConstants.MODID)
@PrefixGameTestTemplate(false)
public class RelayGameTests
{
  private static final String EMPTY_RELAY_TEMPLATE = "relay_activates_from_redstone";

  public RelayGameTests() {}

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void relayActivatesFromRedstone(GameTestHelper helper)
  { RelayTests.relayActivatesFromRedstone(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void invertedRelayRegistersPoweredInput(GameTestHelper helper)
  { RelayTests.invertedRelayRegistersPoweredInput(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void pulseRelayClearsItsPulseAfterTick(GameTestHelper helper)
  { RelayTests.pulseRelayClearsItsPulseAfterTick(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void pulseRelaySetsPoweredFalseWhenInputRemoved(GameTestHelper helper)
  { RelayTests.pulseRelaySetsPoweredFalseWhenInputRemoved(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void bistableRelayLatchesOnFirstRisingEdge(GameTestHelper helper)
  { RelayTests.bistableRelayLatchesOnFirstRisingEdge(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void relayDownFacingRotation0ActivatesFromEast(GameTestHelper helper)
  { RelayTests.relayDownFacingRotation0ActivatesFromEast(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void relayDownFacingRotation0DoesNotActivateFromNorth(GameTestHelper helper)
  { RelayTests.relayDownFacingRotation0DoesNotActivateFromNorth(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void relayDownFacingRotation1ShiftsOutputToEast(GameTestHelper helper)
  { RelayTests.relayDownFacingRotation1ShiftsOutputToEast(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void relayDownFacingRotation1ActivatesFromNorth(GameTestHelper helper)
  { RelayTests.relayDownFacingRotation1ActivatesFromNorth(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void downFacingRotation0FrontIsNorth(GameTestHelper helper)
  { RelayTests.downFacingRotation0FrontIsNorth(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void downFacingRotation1FrontIsEast(GameTestHelper helper)
  { RelayTests.downFacingRotation1FrontIsEast(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void downFacingRotation2FrontIsSouth(GameTestHelper helper)
  { RelayTests.downFacingRotation2FrontIsSouth(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void downFacingRotation3FrontIsWest(GameTestHelper helper)
  { RelayTests.downFacingRotation3FrontIsWest(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void upFacingIsOppositeOfFacingForAllDirections(GameTestHelper helper)
  { RelayTests.upFacingIsOppositeOfFacingForAllDirections(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void downFacingMethodEqualsFacingPropertyForAllDirections(GameTestHelper helper)
  { RelayTests.downFacingMethodEqualsFacingPropertyForAllDirections(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void rightFacingIsRotationPlusOneRelativeToFront(GameTestHelper helper)
  { RelayTests.rightFacingIsRotationPlusOneRelativeToFront(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void backFacingIsRotationPlusTwoRelativeToFront(GameTestHelper helper)
  { RelayTests.backFacingIsRotationPlusTwoRelativeToFront(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void leftFacingIsRotationPlusThreeRelativeToFront(GameTestHelper helper)
  { RelayTests.leftFacingIsRotationPlusThreeRelativeToFront(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void allSixMappedDirectionsAreDistinctForEveryState(GameTestHelper helper)
  { RelayTests.allSixMappedDirectionsAreDistinctForEveryState(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void forwardThenReverseStateMappedFacingIsIdentity(GameTestHelper helper)
  { RelayTests.forwardThenReverseStateMappedFacingIsIdentity(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementDownFaceRotation0(GameTestHelper helper)
  { RelayTests.getStateForPlacementDownFaceRotation0(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementDownFaceRotation1(GameTestHelper helper)
  { RelayTests.getStateForPlacementDownFaceRotation1(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementDownFaceRotation2(GameTestHelper helper)
  { RelayTests.getStateForPlacementDownFaceRotation2(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementDownFaceRotation3(GameTestHelper helper)
  { RelayTests.getStateForPlacementDownFaceRotation3(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementNorthFaceRotation0(GameTestHelper helper)
  { RelayTests.getStateForPlacementNorthFaceRotation0(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementNorthFaceRotation1(GameTestHelper helper)
  { RelayTests.getStateForPlacementNorthFaceRotation1(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementNorthFaceRotation2(GameTestHelper helper)
  { RelayTests.getStateForPlacementNorthFaceRotation2(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementEastFaceRotation1(GameTestHelper helper)
  { RelayTests.getStateForPlacementEastFaceRotation1(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementEastFaceRotation2(GameTestHelper helper)
  { RelayTests.getStateForPlacementEastFaceRotation2(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementSouthFaceRotation1(GameTestHelper helper)
  { RelayTests.getStateForPlacementSouthFaceRotation1(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementSouthFaceRotation3(GameTestHelper helper)
  { RelayTests.getStateForPlacementSouthFaceRotation3(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementWestFaceRotation1(GameTestHelper helper)
  { RelayTests.getStateForPlacementWestFaceRotation1(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementWestFaceRotation3(GameTestHelper helper)
  { RelayTests.getStateForPlacementWestFaceRotation3(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void getStateForPlacementReturnsNullWhenNoSurface(GameTestHelper helper)
  { RelayTests.getStateForPlacementReturnsNullWhenNoSurface(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 10)
  public static void relayDropsWhenSupportBlockRemoved(GameTestHelper helper)
  { RelayTests.relayDropsWhenSupportBlockRemoved(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 30)
  public static void relayDeactivatesWhenPowerRemovedViaTick(GameTestHelper helper)
  { RelayTests.relayDeactivatesWhenPowerRemovedViaTick(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 30)
  public static void bistableRelayDeactivatesOnFallingEdge(GameTestHelper helper)
  { RelayTests.bistableRelayDeactivatesOnFallingEdge(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
  public static void invertedRelayDeactivatesWhenPowerRemoved(GameTestHelper helper)
  { RelayTests.invertedRelayDeactivatesWhenPowerRemoved(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void bridgeRelayPlacedAndGetSignalDoesNotThrow(GameTestHelper helper)
  { RelayTests.bridgeRelayPlacedAndGetSignalDoesNotThrow(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void bridgeRelayGetInputPowerFromRedstoneWire(GameTestHelper helper)
  { RelayTests.bridgeRelayGetInputPowerFromRedstoneWire(helper); }

  @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
  public static void bridgeRelayGetSignalFromLeftWireConnection(GameTestHelper helper)
  { RelayTests.bridgeRelayGetSignalFromLeftWireConnection(helper); }
}

File: fabric/src/main/kotlin/wile/redstonepen/gametest/RelayGameTests.kt
package wile.redstonepen.gametest

import net.minecraft.gametest.framework.GameTest
import net.minecraft.gametest.framework.GameTestHelper
import wile.redstonepen.gametestcommon.RelayTests

@Suppress("UtilityClassWithPublicConstructor")
class RelayGameTests {
    companion object {
        private const val EMPTY_RELAY_TEMPLATE = "redstonepen:relay_activates_from_redstone"

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun relayActivatesFromRedstone(helper: GameTestHelper) =
            RelayTests.relayActivatesFromRedstone(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun invertedRelayRegistersPoweredInput(helper: GameTestHelper) =
            RelayTests.invertedRelayRegistersPoweredInput(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun pulseRelayClearsItsPulseAfterTick(helper: GameTestHelper) =
            RelayTests.pulseRelayClearsItsPulseAfterTick(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun pulseRelaySetsPoweredFalseWhenInputRemoved(helper: GameTestHelper) =
            RelayTests.pulseRelaySetsPoweredFalseWhenInputRemoved(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun bistableRelayLatchesOnFirstRisingEdge(helper: GameTestHelper) =
            RelayTests.bistableRelayLatchesOnFirstRisingEdge(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun relayDownFacingRotation0ActivatesFromEast(helper: GameTestHelper) =
            RelayTests.relayDownFacingRotation0ActivatesFromEast(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun relayDownFacingRotation0DoesNotActivateFromNorth(helper: GameTestHelper) =
            RelayTests.relayDownFacingRotation0DoesNotActivateFromNorth(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun relayDownFacingRotation1ShiftsOutputToEast(helper: GameTestHelper) =
            RelayTests.relayDownFacingRotation1ShiftsOutputToEast(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun relayDownFacingRotation1ActivatesFromNorth(helper: GameTestHelper) =
            RelayTests.relayDownFacingRotation1ActivatesFromNorth(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun downFacingRotation0FrontIsNorth(helper: GameTestHelper) =
            RelayTests.downFacingRotation0FrontIsNorth(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun downFacingRotation1FrontIsEast(helper: GameTestHelper) =
            RelayTests.downFacingRotation1FrontIsEast(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun downFacingRotation2FrontIsSouth(helper: GameTestHelper) =
            RelayTests.downFacingRotation2FrontIsSouth(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun downFacingRotation3FrontIsWest(helper: GameTestHelper) =
            RelayTests.downFacingRotation3FrontIsWest(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun upFacingIsOppositeOfFacingForAllDirections(helper: GameTestHelper) =
            RelayTests.upFacingIsOppositeOfFacingForAllDirections(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun downFacingMethodEqualsFacingPropertyForAllDirections(helper: GameTestHelper) =
            RelayTests.downFacingMethodEqualsFacingPropertyForAllDirections(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun rightFacingIsRotationPlusOneRelativeToFront(helper: GameTestHelper) =
            RelayTests.rightFacingIsRotationPlusOneRelativeToFront(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun backFacingIsRotationPlusTwoRelativeToFront(helper: GameTestHelper) =
            RelayTests.backFacingIsRotationPlusTwoRelativeToFront(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun leftFacingIsRotationPlusThreeRelativeToFront(helper: GameTestHelper) =
            RelayTests.leftFacingIsRotationPlusThreeRelativeToFront(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun allSixMappedDirectionsAreDistinctForEveryState(helper: GameTestHelper) =
            RelayTests.allSixMappedDirectionsAreDistinctForEveryState(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun forwardThenReverseStateMappedFacingIsIdentity(helper: GameTestHelper) =
            RelayTests.forwardThenReverseStateMappedFacingIsIdentity(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementDownFaceRotation0(helper: GameTestHelper) =
            RelayTests.getStateForPlacementDownFaceRotation0(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementDownFaceRotation1(helper: GameTestHelper) =
            RelayTests.getStateForPlacementDownFaceRotation1(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementDownFaceRotation2(helper: GameTestHelper) =
            RelayTests.getStateForPlacementDownFaceRotation2(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementDownFaceRotation3(helper: GameTestHelper) =
            RelayTests.getStateForPlacementDownFaceRotation3(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementNorthFaceRotation0(helper: GameTestHelper) =
            RelayTests.getStateForPlacementNorthFaceRotation0(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementNorthFaceRotation1(helper: GameTestHelper) =
            RelayTests.getStateForPlacementNorthFaceRotation1(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementNorthFaceRotation2(helper: GameTestHelper) =
            RelayTests.getStateForPlacementNorthFaceRotation2(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementEastFaceRotation1(helper: GameTestHelper) =
            RelayTests.getStateForPlacementEastFaceRotation1(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementEastFaceRotation2(helper: GameTestHelper) =
            RelayTests.getStateForPlacementEastFaceRotation2(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementSouthFaceRotation1(helper: GameTestHelper) =
            RelayTests.getStateForPlacementSouthFaceRotation1(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementSouthFaceRotation3(helper: GameTestHelper) =
            RelayTests.getStateForPlacementSouthFaceRotation3(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementWestFaceRotation1(helper: GameTestHelper) =
            RelayTests.getStateForPlacementWestFaceRotation1(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementWestFaceRotation3(helper: GameTestHelper) =
            RelayTests.getStateForPlacementWestFaceRotation3(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun getStateForPlacementReturnsNullWhenNoSurface(helper: GameTestHelper) =
            RelayTests.getStateForPlacementReturnsNullWhenNoSurface(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 10)
        fun relayDropsWhenSupportBlockRemoved(helper: GameTestHelper) =
            RelayTests.relayDropsWhenSupportBlockRemoved(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 30)
        fun relayDeactivatesWhenPowerRemovedViaTick(helper: GameTestHelper) =
            RelayTests.relayDeactivatesWhenPowerRemovedViaTick(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 30)
        fun bistableRelayDeactivatesOnFallingEdge(helper: GameTestHelper) =
            RelayTests.bistableRelayDeactivatesOnFallingEdge(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 20)
        fun invertedRelayDeactivatesWhenPowerRemoved(helper: GameTestHelper) =
            RelayTests.invertedRelayDeactivatesWhenPowerRemoved(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun bridgeRelayPlacedAndGetSignalDoesNotThrow(helper: GameTestHelper) =
            RelayTests.bridgeRelayPlacedAndGetSignalDoesNotThrow(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun bridgeRelayGetInputPowerFromRedstoneWire(helper: GameTestHelper) =
            RelayTests.bridgeRelayGetInputPowerFromRedstoneWire(helper)

        @JvmStatic
        @GameTest(template = EMPTY_RELAY_TEMPLATE, timeoutTicks = 5)
        fun bridgeRelayGetSignalFromLeftWireConnection(helper: GameTestHelper) =
            RelayTests.bridgeRelayGetSignalFromLeftWireConnection(helper)
    }
}
